Working on an IP based connection limit to harden APIServer.

cheftako opened the pull request:
Implemented a chain handler for handling the connection options.
(keep-alive, ip-limiting, ...)
Added white list functionality for loopback.
Added a total count of connections limit as well.
Put the IpLimit into ServingInfo and set up via config options.
Added tests to cover the basic behaviors of IpBaseLimit functionality.
Added parrallel access tests for IPBasedLimit
Putting in the generated files.

**Release note**:
Formerly the APIServer only limited requests after the SSL handshake. 
This fix is not applied to localhost or non SSL connections. 
The limit is applied on both a total and per IP basis.
The IP limit is the smaller of either half the write or read limit.
The total limit is the sum of the write and read limits.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39901)
<!-- Reviewable:end -->

cheftako commented:
I signed it

cheftako commented:
I signed it

cheftako committed:
Working on an IP based connection limit to harden APIServer.

Implemented a chain handler for handling the connection options.
(keep-alive, ip-limiting, ...)
Added white list functionality for loopback.
Added a total count of connections limit as well.
Put the IpLimit into ServingInfo and set up via config options.
Added tests to cover the basic behaviors of IpBaseLimit functionality.
Added parrallel access tests for IPBasedLimit
Putting in the generated files.
Refactored the code to use server conn status change not conn wrapper.
Fixed minor issue with uncomsumed data before close.
Added debug library to help with issues like above.
Added improved debug for connections and fixed hijack bug.
Fixed to failing unit tests.
First test was a result of the new DebugConcurrent flag.
The second was a change in the YAML to JSON error message.
Seeing a better error message which broke the test.
The error is now prepended with info that its a YAML to JSON error.
Can now hit http://localhost:8080/debug/connections/secure for info.
Http StateHijack is terminal and is now treated that way in conn mgmt.
Fixed config to use alternate no secure connection method for debug when
not configed.
Fixed serve_test to initialize decrementer for IPBasedLimit.
Switched debug/connections to be JSON rather than plain text.
Fixes suggested by lavalamp.
Injected missing line from merge.
Fixed Ip to be IP
Make sure to properly lock Decrementers.
Added command line configuration of max connections.

cheftako commented:
@k8s-bot gce etcd3 e2e test this

cheftako commented:
@k8s-bot gci gce e2e test this

cheftako commented:
@k8s-bot node e2e test this

cheftako commented:
@k8s-bot kops aws e2e test this

cheftako commented:
@k8s-bot cvm gce e2e test this

lavalamp commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
-	tc.SetKeepAlivePeriod(defaultKeepAlivePeriod)
-	return tc, nil
+	local := conn.LocalAddr().String()
+	glog.Infof("Entered ip based limit Accept() for IP %s on %s", ip, local)
```
remove (or .V(9)).

lavalamp commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	err = h.ipLimit.incrementIp(ip)
+	if err != nil {
+		conn.Close() // Not decrementing as we did not successfully increment
+		glog.Infof("Denied accept to IP %s", ip)
```
.V(2), maybe?

Also rephrase so a random sysadmin reading the logs will not open an issue / send us mail asking what happened :)

("Blocked connection from %v, as it has exceeded its connection count. Read the help for --flag-name if this seems wrong." ... or some such)

liggitt replied:
so this does a straight connection close, with no http or API response? that seems wrong. have we verified what control plane components do when they get this response? there is existing logic (though I'm not sure how well tested) that tries to behave well in the presence of 429 errors, but I'm not sure what clients would do with this

cheftako replied:
The goal here is to limit concurrent connection prior to taking on the expense of having completed (or even started) the SSL handshake. Given that I don't think there is a practical way to return a usable HTTP or API response.

liggitt replied:
Hmm, not sure how I feel about breaking the API in this case. I'd want to be super sure we had the load balancer and non-127.0.0.1 API server paths well tested to make sure anything that hit this was actually likely to be abuse/flooding

lavalamp commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+		glog.Infof("Denied accept to IP %s", ip)
+		return nil, err
+	}
+	glog.Infof("Allowed accept to IP %s", ip)
```
remove?

lavalamp commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+
+func (c *ipBasedLimitListenerConn) Close() error {
+	err := c.Conn.Close()
+	c.releaseOnce.Do(func() { c.ipLimit.decrementIp(c.ip) })
```
Not claiming this is necessarily better but an option to consider is:

```go
defer c.releaseOnce.Do(func() { c.ipLimit.decrementIp(c.ip) })
return c.Conn.Close()
```

cheftako replied:
Seems better to me.

lavalamp commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+
+// A custom error which tells http/server.go Serve() to retry Accept() quickly
+type limitedError struct {
+	// Present for debugging - IP which saw the problem
```
Seems reasonable for error reporting, not just debugging.

lavalamp commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+}
+
+func (l limitedError) Error() string {
+	return "Accept() Exceeded maximum concurrent requests for IP " + l.ip
```
Or exceeded the overall limit?

Go error messages traditionally start with lowercase letters and don't end with punctuation (so that they can be concatenated and not look weird if there are multiple ones). That's a long way of saying s/Exceeded/exceeded/.

lavalamp commented on pkg/genericapiserver/server/serve_test.go:
```diff
@@ -564,6 +566,312 @@ NextTest:
 }
 
+func TestBasicIpLimit(t *testing.T) {
+	var ipLimit ipBasedLimit
```
I think this test would work well as a table. See https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go

lavalamp commented on pkg/genericapiserver/server/serve_test.go:
```diff
@@ -564,6 +566,312 @@ NextTest:
 	}
 }
 
+func TestBasicIpLimit(t *testing.T) {
```
s/Ip/IP/g

lavalamp commented on pkg/genericapiserver/server/serve_test.go:
```diff
@@ -564,6 +566,312 @@ NextTest:
+	assert.EqualValues(t, val, 0, "Deleting decrement call for ipBasedLimit has a val of %d", 0)
+}
+
+func TestParrallelIpLimitAccessNoLimit(t *testing.T) {
```
s/Parrallel/Parallel/g :)

s/parrallel/parallel/g too.

lavalamp commented on pkg/genericapiserver/server/serve_test.go:
```diff
@@ -564,6 +566,312 @@ NextTest:
+	var ipLimit ipBasedLimit
+	validateIpLimitState(t, &ipLimit, "Basic/raw IpBasedLimit", 0, 0, 0, 0)
+	ipLimit.incrementIp("1.2.3.4")
+	validateIpLimitState(t, &ipLimit, "Basic/raw IpBasedLimit increment", 0, 0, 0, 0)
```
wouldn't the map size be 1 at this point?

lavalamp replied:
oh, I see, this is configured to block everything. nvm.

lavalamp commented on pkg/genericapiserver/server/serve_test.go:
```diff
@@ -564,6 +566,312 @@ NextTest:
+
+func TestParrallelIpLimitAccessNoLimit(t *testing.T) {
+	ipLimit := ipBasedLimit{limits: make(map[string]int), ipMax: 10, max: 100}
+	stop := time.Now().Add(time.Second * 10) // Long enough to prevent flaky ???
```
Consider making this just one second, and reducing the other time from 100ms to 10ms, just to keep the test wall time down?

Also consider instead of using a stop time using a fixed number of iterations--a stop time means it may be more or less well tested depending on how much CPU contention there is. We have a bit of a flaky test problem so it'd be good not to make it worse :)

lavalamp commented on pkg/genericapiserver/server/serve_test.go:
```diff
@@ -564,6 +566,312 @@ NextTest:
+	ips := []string{"1.1.1.1", "2.2.2.2", "3.3.3.3", "4.4.4.4", "5.5.5.5",
+		"6.6.6.6", "7.7.7.7", "8.8.8.8", "9.9.9.9",
+		"1973:dead:beef:0121:1973:dead:beef:0121"}
+	done := make(chan struct{})
```
Insert a `,100`-- as written (unbuffered), the channel is acting to serialize all operations done on it.

lavalamp commented on pkg/genericapiserver/server/serve_test.go:
```diff
@@ -564,6 +566,312 @@ NextTest:
+	ips := []string{"1.1.1.1", "2.2.2.2", "3.3.3.3", "4.4.4.4", "5.5.5.5",
+		"6.6.6.6", "7.7.7.7", "8.8.8.8", "9.9.9.9",
+		"1973:dead:beef:0121:1973:dead:beef:0121"}
+	done := make(chan struct{})
```
Same comments as previous test :)

lavalamp reviewed (commented):
No major problems, just a bunch of little things. Thanks!

lavalamp commented:
Judging from the e2e test failures, it seems the tests themselves are getting rate limited by this?

cheftako commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -93,13 +93,13 @@ func (s *GenericAPIServer) serveInsecurely(stopCh <-chan struct{}) error {
 	glog.Infof("Serving insecurely on %s", s.InsecureServingInfo.BindAddress)
 	var err error
-	s.effectiveInsecurePort, err = runServer(insecureServer, s.InsecureServingInfo.BindNetwork, stopCh)
+	s.effectiveInsecurePort, err = runServer(insecureServer, s.InsecureServingInfo.BindNetwork, &s.InsecureServingInfo.ipLimit, stopCh)
```
It is effectively not being turned on for the insecure port. The InsecureServingInfo is never really initialized (its all defaults). The default behavior is wide open with no limits applied.

mml commented on pkg/genericapiserver/server/config.go:
```diff
@@ -169,6 +169,9 @@ type ServingInfo struct {
 	// "tcp4", and "tcp6".
 	BindNetwork string
+
+	// ipBasedLimit determines the maximum number of concurrent requests based on ip
```
The comment should refer to the field name rather than the type name.

liggitt replied:
everything in this struct should be exported. this config is what allows driving the generic api server from consuming components

mml commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
 
-func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
-	tc, err := ln.AcceptTCP()
+// ipBasedLimitListener accepts at most n simultaneous connections from a
```
This comment confuses me.  There isn't a type named ipBasedLimitListener.  Does it belong with ipBasedLimitListenerConn?

cheftako replied:
It should be ipBasedLimitHandler. I failed to fix the comment when refactoring. Good catch.

mml commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	mutex sync.Mutex
+
+	// The maximum concurent connections an IP is allowed to have open
+	ipMax int
```
naming nit: `maxPerIP` or even `maxPerAddress` might be clearer.

liggitt replied:
is a uniform limit going to be sufficient? it seems likely that external traffic will come through a load balancer and appear as a single IP

deads2k commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -134,7 +134,11 @@ func runServer(server *http.Server, network string, stopCh <-chan struct{}) (int
+				tcpKeepAliveHandler{},
+				ipBasedLimitHandler{ipLimit},
+			}
+			listener = tcpConnectionChainListener{ln.(*net.TCPListener), chain}
```
@liggitt @sttts FYI, you guys have been down this deep.

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
 	return byName, nil
 }
 
+// tcpConnectionChainListener is responsible to invoking any connection handling
```
Can these helpers go to something like iplimiter.go? config.go is not really the right place.

sttts replied:
misread the patch. We are in serve.go already.

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
-	tc, err := ln.AcceptTCP()
+// ipBasedLimitListener accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
```
Be aware of non-0.0.0.0 bind addresses, compare https://github.com/kubernetes/kubernetes/commit/5b1d45bc150f7e007cf842bcab08344e92a8ed38#diff-d6d404f0eeafae80d69923ebbd3880d5R52. I.e. the loopback client sometimes does not talk to 127.0.0.1, but to some interface address.

cheftako replied:
The IP being checked here is remote address (from the servers perspective). It seems like that should always have been resolved to an actual IP address. We are also determining local using the IsLoopback method from IP. (Sort of feels like these sort of cases should be handled there rather than here) @sttts  would love to get your thoughts on this.

sttts replied:
My guess is that the remote address won't be 127.0.0.1 either. Please double check.

cheftako replied:
So in my testing the majority of IP address I see are either "127.0.0.1" or "::1" (Eg. "Allowed accept to IP 127.0.0.1" and "Allowed accept to IP ::1") I assume this is all the parts of the master talking to the api-server. So I know we do see those addresses. I have yet to see a "0.0.0.0" address in any of my testing but if you have a suggestion on how to generate it I would be happy to give it a try.

sttts replied:
0.0.0.0 leads to 127.0.0.1 for loopback client as target IP. You have to bind to the IP of one of your interfaces, e.g. 192.168.2.42 if that is the eth0 IP. Then the apiserver will use that one to connect to itself. Then check what the source address is.

liggitt replied:
Did this get addressed?

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	total int
+
+	// Lock used to ensure the limits and total are updated atomicaly
+	mutex sync.Mutex
```
convention: put that above the protected vars.

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	i.total--
+	current := i.limits[ipser]
```
Do we have a programming error if `ipset` does not exist anymore? Would be good to detect that. Maybe a warning would make sense here instead of in the normal case as @lavalamp commented correctly.

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	return nil
+}
+
+func (i *ipBasedLimit) decrementIp(ipser string) {
```
incrementIp and decrementIp look nearly symmetric. Maybe merge them?

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+// assume if an error is returned from a chain handler that it will return a nil
+// conn and have already closed that conn.
+type tcpConnectionChainHandler interface {
+	init(*net.TCPConn) error
```
even though this interface is only used locally, I would still use capital-case method names to make clear in the implementation that these belong to some API.

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+}
+
+// A custom error which tells http/server.go Serve() to retry Accept() quickly
+type limitedError struct {
```
a bit more verbosity would make it easier to read: e.g. IpLimitExceededError

liggitt commented:
> The IP limit is the smaller of either half the write or read limit.
> The total limit is the sum of the write and read limits.

don't forget long-running request limits, which are distinct

cheftako commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+type ipBasedLimit struct {
+	// Map in which we map the serialized ip to the current count
```
They do but then we have to worry about IPv4 vs IPv6. I have switched it over to plain IP.

cheftako replied:
Just realized why I used a string. IP is a slice and you cannot use a slice as the key in a map.

sttts commented on pkg/genericapiserver/server/serve.go:
```diff
@@ -78,7 +78,7 @@ func (s *GenericAPIServer) serveSecurely(stopCh <-chan struct{}) error {
 	glog.Infof("Serving securely on %s", s.SecureServingInfo.BindAddress)
 	var err error
-	s.effectiveSecurePort, err = runServer(secureServer, s.SecureServingInfo.BindNetwork, stopCh)
+	s.effectiveSecurePort, err = runServer(secureServer, s.SecureServingInfo.BindNetwork, &s.SecureServingInfo.ipLimit, stopCh)
```
Not the SecureServingInfo, but ServingInfo should be fine.

cheftako commented:
@k8s-bot Jenkins verification test this

cheftako commented:
@k8s-bot Jenkins unit/Integration test this

cheftako commented:
@ixdy @fejta The link the build system provided doesn't seem to work (https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39901/pull-kubernetes-unit/15549/) Any clues or ideas?

fejta commented:
Problem was an infrastructure failure,  flake connecting to github, so we even get to the point where anything started (including the program that uploads results)

```
+ git clone https://github.com/kubernetes/test-infra -b master
Cloning into 'test-infra'...
fatal: unable to access 'https://github.com/kubernetes/test-infra/': Failed to connect to github.com port 443: Connection timed out
```

cheftako commented:
@k8s-bot GCE etcd3 e2e test this
@k8s-bot kops AWS e2e test this

cheftako commented:
@k8s-bot GCE etcd3 e2e test this

cheftako commented:
@k8s-bot kops aws e2e test this

cheftako commented:
@k8s-bot gce etcd3 e2e test this

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -93,23 +95,31 @@ func (s *GenericAPIServer) serveInsecurely(stopCh <-chan struct{}) error {
+	if len(info.BindNetwork) == 0 {
+		info.BindNetwork = "tcp"
+	}
+	server.ConnState = func(conn net.Conn, state http.ConnState) {
```
Would it make sense to make this a function on the IP limiter? So this line would read `server.ConnState = info.IPLimit.ConnState`. Might be a little clearer.

cheftako replied:
ConnState seems a bit vague to me. So I called the IPBasedLimit method HandleHttpServerStateChange

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -93,23 +95,31 @@ func (s *GenericAPIServer) serveInsecurely(stopCh <-chan struct{}) error {
+		info.BindNetwork = "tcp"
+	}
+	server.ConnState = func(conn net.Conn, state http.ConnState) {
+		if state == http.StateClosed || state == http.StateHijacked {
```
Websockets, exec, attach, portforward, etc all get hijacked. So effectively those connections won't count against the limit after they've been accepted. Not necessarily a problem, just pointing this out.

cheftako replied:
In fact StateHijacked is a terminal state in the http server code, so they specifically must not count against the limit. That's why this code decrements when the connection goes in to the hijack state.

liggitt replied:
I don't think logs and watch hijack, and those are considered long-running

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -93,23 +95,31 @@ func (s *GenericAPIServer) serveInsecurely(stopCh <-chan struct{}) error {
+	}
+	server.ConnState = func(conn net.Conn, state http.ConnState) {
+		if state == http.StateClosed || state == http.StateHijacked {
+			decrementer := info.IPLimit.decrementers[conn.RemoteAddr().String()]
```
Optional nit, can compress to one line.

```go
if dec, ok := mapname[key]; ok { // assumes you don't store nils in the map
  dec()
}
```

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+}
+
+func NoSecureHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "{ \"Error\": \"No secure server configured\" }")
```
You can use backticks to quote and then you don't have to escape the quotes.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	encoder.SetIndent("", "\t")
+	err := encoder.Encode(&s.IPLimit)
+	if err != nil {
+		// glog.V(2).Infof("Error marshaling secure server info: %v", err)
```
remove dead code

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	fmt.Fprintf(w, "{ \"Error\": \"No secure server configured\" }")
+}
+
+// tcpConnectionChainListener is responsible to invoking any connection handling
```
s/Listener/Handler/

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+// TODO: (cheftako) : IPv6, CIDR Range rather than individual IP Addresses?
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+
```
remove blank line (I guess not important here, but for public types a blank line here will break godoc).

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	ip := net.ParseIP(sourceIP)
+	if ip == nil {
+		glog.Infof("Received request from client with invalid ip %s", sourceIP)
+		return IPLimitExceededError{ip: sourceIP, reason: "unparseable ip address"}
```
Maybe append the ip address to this?

cheftako replied:
Assuming you are referring to the error, the ip will be added when the actual Error() message is generated. The full Error() message should look like 
"Accept() blocked connection request from 1.2.3.4 because unparseable ip address."

lavalamp replied:
SGTM

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+	defer i.mutex.Unlock()
+	current := i.limits[sourceIP]
+	if current >= i.maxPerIP {
+		return IPLimitExceededError{ip: sourceIP, reason: fmt.Sprintf(" exceeded per IP limit of %d", i.maxPerIP)}
```
Leading space seems unnecessary. (below, too.)

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+// A) It allows us to minimize the scope of the mutex which needs to be held
+// B) It allows the fields to remain private
+// C) It allows us to marshal 'decrementers' as a list of its keys
+// This is only meant for use in the debug API.
```
Interesting approach-- in the past I've just declared a type specifically to be marshalled, written code to fill it, and then use the standard JSON package to actually marshal it. Less chance of bugs :)

cheftako replied:
I thought about that. However to make it work I would need to essentially use the logic I have here to safely copy the data into the new struct, while minimizing the mutex scope and converting the map to an actual list. It felt like I was trading less formatting code (braces and commas) for more memory churn and more lines of code (the additional struct).

lavalamp replied:
Given that it's just debug code, I'd personally optimize for code that's obviously got no bugs :) But I think memory churn is a reasonable point and I don't feel super strongly.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
+// This is only meant for use in the debug API.
+func (i *ipBasedLimit) MarshalJSON() ([]byte, error) {
+	buffer := bytes.NewBufferString("{")
+	buffer.WriteString(fmt.Sprintf("\"MaxPerIP\": %d, ", i.maxPerIP))
```
better: `fmt.Fprintf(buffer, "format string", args)`

lavalamp replied:
Also, backticks will make the quoting more readable. (but also mean that \n wouldn't be a newline, if that matters.)

lavalamp commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -114,6 +114,9 @@ type Config struct {
 	// The maximum length of time to wait before giving up on a server request. A value of zero means no timeout.
 	Timeout time.Duration
 
+	// Flag to indicate whether we want to enable the fairly heavy weight debug concurrent functionality
```
Add (or reference) description of why one might want to do that. :)

lavalamp commented on staging/src/k8s.io/client-go/transport/config.go:
```diff
@@ -80,6 +84,11 @@ func (c *Config) HasCertAuth() bool {
 	return len(c.TLS.CertData) != 0 || len(c.TLS.CertFile) != 0
 }
 
+// IsDebugConcurrent returns whether we should be debugging concurrent state.
```
Go accessors are usually of the form "X" as opposed to "GetX" or "IsX". Since DebugConcurrent is already public, do you really need this method?

cheftako replied:
Tempted to suggest we head more toward accessors and better obfuscation. However as most of the Config is already public not sure its worth worrying about.

lavalamp commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -261,6 +268,54 @@ func (rt *impersonatingRoundTripper) CancelRequest(req *http.Request) {
+// open transports and provide that information if something goes wrong.
+
+var debugConcurrentMutex sync.Mutex
+var debugConcurrentMap map[string]debugConcurrentRecord = make(map[string]debugConcurrentRecord)
```
I would prefer to wrap these in an object. Even if there's a singleton instance of the object, at least that way you could declare a second one if you wanted to (for testing, say).

lavalamp reviewed (commented):
Some more minor change requests. Looking pretty good!

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -260,9 +263,15 @@ func (c *Config) ApplySecureServingOptions(secureServing *options.SecureServingO
+	if ipMax > c.MaxMutatingRequestsInFlight/2 {
+		ipMax = c.MaxMutatingRequestsInFlight / 2
+	}
+	max := c.MaxRequestsInFlight + c.MaxMutatingRequestsInFlight
```
The only long running feature I am aware of is --long-running-request-regexp "which should be excluded from maximum inflight request handling". So that bucket does not have a pool size per se. It seemed simplest to me  to claim this was not part of "max inflight request handling". In part I was advocating for this as at the time we are enforcing the ipLimit we don't know what the URL will be for the regexp match. If we decide to do it, we will probably need to fix the regexp code to call the decrementer on a match.

liggitt replied:
This seems unresolved.

cheftako replied:
Hi @liggitt what were you looking for here?

liggitt replied:
> It seemed simplest to me to claim this was not part of "max inflight request handling"

they consume some of the connection count, right?

cheftako replied:
Yes, they consume some of the maximum connection count. However they still do not count against the maximum number of in-flight requests. So the documentation is still technically correct. Do you feel like we should not count requests matching the regex from counting against the connection limit? This is actually fairly hairy to handle. Not only would we need to worry about decrementing the limit when we got a match, (make sure we do not over-decrement if we allow multiple requests per connection) but we would also then need to worry about re-incrementing the count once that request was finished.

lavalamp replied:
Hm, it's a little weird that those requests are *completely* unlimited. That seems like an oversight. We should probably limit them.

In the interest of making progress, how about we add some generous margin to this limit in this PR to account for them, and then in a follow-up set an explicit limit on long-running requests?

liggitt replied:
Agree we shouldn't have any completely unbounded counts, I just don't have a good sense of the number of long running connections we have today (for example, in the 5k kubemark tests), so I'm not sure what to set the initial limit at

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/routes/profiling.go:
```diff
@@ -26,8 +27,9 @@ import (
 
 // Install adds the Profiling webservice to the given mux.
-func (d Profiling) Install(c *mux.APIContainer) {
+func (d Profiling) Install(c *mux.APIContainer, secureConnHandler func(http.ResponseWriter, *http.Request)) {
```
Agreed. Should probably be broken out of profiling as well (into something like debug) and given its own control flag.

lavalamp commented:
/lgtm
/approve

liggitt commented:
Needs squash, and I didn't see the comment about the majority of traffic arriving via a load balancer

I also am concerned about the API *sometimes* indicating maximum requests have been exceeded by dropping the connection entirely. At the very least, that seems like a behavior you should need to opt into for your API.

lavalamp commented:
> concerned about the API sometimes indicating maximum requests have been exceeded by dropping the connection entirely. At the very least, that seems like a behavior you should need to opt into for your API.

The pre-SSL handshake limit is a safety valve. It should be slightly more permissive than the post-SSL handshake limits, such that usually when people exceed the limit, they get a nice error, but if they exceed the limit by a great number of requests, apiserver defensively sheds that load. I think it's totally OK for someone sending that much load to get abrupt connection closures, and if the limits are set up like I suggest, it's also true that they'll see some scattered helpful error messages.

I would really like to avoid adding a flag that we expect sysadmins to do something with. I can see adding a flag if we're worried about bugs and want an escape hatch, but given that you can already adjust the limits, I would still prefer to not add that, either.

cheftako commented:
@k8s-bot bazel test this

cheftako commented:
@k8s-bot gci gke e2e test this

cheftako commented:
@k8s-bot bazel test this

cheftako commented:
@k8s-bot unit test this

cheftako commented:
@k8s-bot kops aws e2e test this

cheftako commented:
@k8s-bot cvm gke e2e test this

cheftako commented:
@k8s-bot cvm gce e2e test this

cheftako commented:
@k8s-bot gci gce e2e test this

cheftako commented:
@k8s-bot kubemark e2e test this

cheftako commented:
@k8s-bot gce etcd3 e2e test this

cheftako commented:
@k8s-bot bazel test this

cheftako commented:
@k8s-bot cvm gke e2e test this

cheftako commented:
@k8s-bot verify test this

cheftako commented:
@k8s-bot cross build this

liggitt commented on cluster/gce/config-default.sh:
```diff
@@ -24,8 +24,8 @@ ZONE=${KUBE_GCE_ZONE:-us-central1-b}
-NODE_SIZE=${NODE_SIZE:-n1-standard-2}
-NUM_NODES=${NUM_NODES:-3}
+NODE_SIZE=${NODE_SIZE:-n1-standard-32}
+NUM_NODES=${NUM_NODES:-16}
```
revert change

cheftako replied:
Will do, this was part of the kubemark tests you had suggested and will not be pushed.

liggitt replied:
ah. the full-scale kubemark test doesn't run on the merge queue, as far as I know. @wojtek-t, can you confirm?

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/routes/profiling.go:
```diff
@@ -26,8 +27,9 @@ import (
 	c.UnlistedRoutes.HandleFunc("/debug/pprof/", pprof.Index)
 	c.UnlistedRoutes.HandleFunc("/debug/pprof/profile", pprof.Profile)
 	c.UnlistedRoutes.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+	c.UnlistedRoutes.HandleFunc("/debug/connections/secure", secureConnHandler)
```
cc @smarterclayton I know you were looking at exposing debug endpoints

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/serving.go:
```diff
@@ -165,9 +165,16 @@ func (s *SecureServingOptions) applyServingInfoTo(c *server.Config) error {
+	if ipMax > c.MaxMutatingRequestsInFlight/2 {
+		ipMax = c.MaxMutatingRequestsInFlight / 2
+	}
+	max := c.MaxRequestsInFlight + c.MaxMutatingRequestsInFlight
```
again, this max doesn't take into account long-running requests (which don't currently fall into either the maxrequests or maxmutatingrequests buckets)

cheftako replied:
We've been trying to limit the need to expose additional config parameters. For small masters (1-4 CPU on master and a limited number of nodes) these values seem to be correct and provide much better resilience. However I think with differing sizes and deployment types, the sort of config option you suggested is probably the best choice.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +231,222 @@ func GetNamedCertificateMap(certs []NamedTLSCert) (map[string]*tls.Certificate,
+	i.total++
+	i.Limits[sourceIP] = current + 1
+	var releaseOnce sync.Once
+	i.Decrementers[conn] = func() {
```
I see us locking here to write to Decrementers, but not locking to read in HandleHttpServerStateChange... how does that not cause races?

cheftako replied:
It does as I demonstrated yesterday when running the kubemark tests you suggested. I have fixed this locally but am still trying to address your main concern which prompted the kubemark tests.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/serve.go:
```diff
@@ -224,22 +231,222 @@ func GetNamedCertificateMap(certs []NamedTLSCert) (map[string]*tls.Certificate,
+// B) It allows the fields to remain private
+// C) It allows us to marshal 'Decrementers' as a list of its keys
+// This is only meant for use in the debug API.
+func (i *IPBasedLimit) MarshalJSON() ([]byte, error) {
```
wouldn't these be better exposed as prometheus metrics?

smarterclayton commented:
So every large Kubernetes cluster in aware of is behind a load balancer and source ip is not the origin.  I *assume* that with this change those clusters have to set some new value?  

I think we see 20k active API connects from our load balancers under high load in large clusters, although http2 has dropped.  We also have lots of web consumers, each of whom has multiple simultaneous watches (we size for 2-3k watches from web console users).  What's the impact to that?

smarterclayton commented:
@kubernetes/sig-scalability-pr-reviews for a quick look (unless they've already been in this pr)

cheftako commented:
@k8s-bot cvm gke e2e test this

cheftako commented:
@k8s-bot gci gke e2e test this
